File: src/main/java/org/example/tree/TreeDiameter.java
package org.example.tree;

public class TreeDiameter {

    private int maxD = 0;
    public int diameterOfBinaryTree(TreeNode root) {
        if (root == null) return 0;

        dfs(root);

        return maxD;
    }

    private int dfs(TreeNode node) {
        if (node == null) return 0;

        int left = dfs(node.left);
        int right = dfs(node.right);

        // Diameter = (maxL + 1) + (maxR + 1)
        maxD = Math.max(maxD, left + right);

        // this calculates height
        return 1 + Math.max(left, right);
    }
}

File: src/main/java/org/example/tree/TreeDiameter.py
from typing import Optional

from org.example.tree.TreeNode import TreeNode


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = [0]

        def dfs(root):
            if not root: return 0

            left = dfs(root.left)
            right = dfs(root.right)
            res[0] = max(res[0], left + right)  # diameter from a particular node

            return 1 + max(left, right)  # tree height

        dfs(root)
        return res[0]
